ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,402 @@
+impl CurrentTime for SystemTime {
+	fn timestamp(&self) -> u64 {
+		self.duration_since(UNIX_EPOCH).unwrap_or_default().as_secs()
+	}
+}
+
+/// Private transactions logging
+pub struct Logging {
+	logs: RwLock<HashMap<H256, TransactionLog>>,
+	logs_serializer: Arc<LogsSerializer>,
+	timestamp_source: Box<CurrentTime>,
+}
+
+impl Logging {
+	/// Creates the logging object
+	pub fn new(logs_serializer: Arc<LogsSerializer>, timestamp_source: Box<CurrentTime>) -> Self {
+		let logging = Logging {
+			logs: RwLock::new(HashMap::new()),
+			logs_serializer,
+			timestamp_source,
+		};
+		if let Err(err) = logging.read_logs() {
+			warn!(target: "privatetx", "Cannot read logs: {:?}", err);
+		}
+		logging
+	}
+
+	/// Retrieves log for the corresponding tx hash
+	pub fn tx_log(&self, tx_hash: &H256) -> Option<TransactionLog> {
+		self.logs.read().get(&tx_hash).cloned()
+	}
+
+	/// Logs the creation of the private transaction
+	pub fn private_tx_created(&self, tx_hash: &H256, validators: &[Address]) {
+		let mut validator_logs = Vec::new();
+		for account in validators {
+			validator_logs.push(ValidatorLog {
+				account: *account,
+				validation_timestamp: None,
+			});
+		}
+		let mut logs = self.logs.write();
+		if logs.len() > MAX_JOURNAL_LEN {
+			// Remove the oldest log
+			if let Some(tx_hash) = logs.values()
+				.min_by(|x, y| x.creation_timestamp.cmp(&y.creation_timestamp))
+				.map(|oldest| oldest.tx_hash)
+			{
+				logs.remove(&tx_hash);
+			}
+		}
+		logs.insert(*tx_hash, TransactionLog {
+			tx_hash: *tx_hash,
+			status: PrivateTxStatus::Created,
+			creation_timestamp: self.timestamp_source.timestamp(),
+			validators: validator_logs,
+			deployment_timestamp: None,
+			public_tx_hash: None,
+		});
+	}
+
+	/// Logs the validation of the private transaction by one of its validators
+	pub fn signature_added(&self, tx_hash: &H256, validator: &Address) {
+		let mut logs = self.logs.write();
+		if let Some(transaction_log) = logs.get_mut(&tx_hash) {
+			if let Some(ref mut validator_log) = transaction_log.validators.iter_mut().find(|log| log.account == *validator) {
+				transaction_log.status = PrivateTxStatus::Validating;
+				validator_log.validation_timestamp = Some(self.timestamp_source.timestamp());
+			}
+		}
+	}
+
+	/// Logs the final deployment of the resulting public transaction
+	pub fn tx_deployed(&self, tx_hash: &H256, public_tx_hash: &H256) {
+		let mut logs = self.logs.write();
+		if let Some(log) = logs.get_mut(&tx_hash) {
+			log.status = PrivateTxStatus::Deployed;
+			log.deployment_timestamp = Some(self.timestamp_source.timestamp());
+			log.public_tx_hash = Some(*public_tx_hash);
+		}
+	}
+
+	fn read_logs(&self) -> Result<(), Error> {
+		let mut transaction_logs = self.logs_serializer.read_logs()?;
+		// Drop old logs
+		let current_timestamp = self.timestamp_source.timestamp();
+		transaction_logs.retain(|tx_log| current_timestamp - tx_log.creation_timestamp < MAX_STORING_TIME);
```

Comment:
Looking at the `CurrentTime` implementation for `SystemTime` [above](https://github.com/paritytech/parity-ethereum/pull/10056/files#diff-bc698921b3e50488ef5d7b2875e5436bR163) it seems that errors are mapped to 0. Assuming `tx_log.creation_timestamp > 0` this could lead to an underflow here (or a panic in `debug` mode).